### ISSUE TITLE
Crash after Changing due date

### ISSUE THREAD

**Issue by stackcoder** [dmfs/opentasks#823]
Steps to reproduce:
- Create a new task and set due
- Save task
- Edit task and update due date
- Press `SAVE`
- App crashes and looses all changes

```
08-01 16:20:32.583  7833  7833 D Profiled: Time spent in Update: 2 milliseconds
08-01 16:20:32.584  7833  7833 D AndroidRuntime: Shutting down VM
08-01 16:20:32.585  7833  7833 E AndroidRuntime: FATAL EXCEPTION: main
08-01 16:20:32.585  7833  7833 E AndroidRuntime: Process: org.dmfs.tasks, PID: 7833
08-01 16:20:32.585  7833  7833 E AndroidRuntime: java.lang.IllegalArgumentException: _ID can not be set manually
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.processors.tasks.Validating.verifyCommon(Validating.java:57)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.processors.tasks.Validating.update(Validating.java:2)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.processors.tasks.Validating.update(Validating.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.TaskProvider.updateInTransaction(TaskProvider.java:20)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.SQLiteContentProvider.a(SQLiteContentProvider.java:28)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.d.value(Unknown Source:10)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.utils.Profiled.run(Profiled.java:5)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.SQLiteContentProvider.update(SQLiteContentProvider.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.provider.tasks.TaskProvider.update(TaskProvider.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.content.ContentProvider$Transport.update(ContentProvider.java:361)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.content.ContentResolver.update(ContentResolver.java:1731)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.tasks.model.ContentSet.persist(ContentSet.java:5)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.tasks.EditTaskFragment.saveAndExit(EditTaskFragment.java:27)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at org.dmfs.tasks.EditTaskFragment.onOptionsItemSelected(EditTaskFragment.java:2)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.fragment.app.Fragment.performOptionsItemSelected(Fragment.java:3)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.fragment.app.FragmentManagerImpl.dispatchOptionsItemSelected(FragmentManager.java:4)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.fragment.app.FragmentController.dispatchOptionsItemSelected(FragmentController.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.fragment.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:3)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.app.AppCompatDelegateImpl.onMenuItemSelected(AppCompatDelegateImpl.java:4)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:2)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:4)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.widget.ActionMenuView.invokeItem(ActionMenuView.java:1)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at androidx.appcompat.view.menu.ActionMenuItemView.onClick(ActionMenuItemView.java:2)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.view.View.performClick(View.java:6617)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.view.View.performClickInternal(View.java:6594)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.view.View.access$3100(View.java:778)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.view.View$PerformClick.run(View.java:25910)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:193)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6739)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:495)
08-01 16:20:32.585  7833  7833 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:859)
08-01 16:20:32.587  3405 10602 W ActivityManager:   Force finishing activity org.dmfs.tasks/.EditTaskActivity
08-01 16:20:32.592  3405 10602 W ContextImpl: Calling a method in the system process without a qualified user: android.app.ContextImpl.sendBroadcast:1005 com.android.server.am.AppErrors.crashApplicationInner:469 com.android.server.am.AppErrors.crashApplication:392 com.android.server.am.ActivityManagerService.handleApplicationCrashInner:15762 com.android.server.am.ActivityManagerService.handleApplicationCrash:15728 
08-01 16:20:32.593  7833  7833 I Process : Sending signal. PID: 7833 SIG: 9
```

**Comment by dmfs** [comment]
Apparently the UI tries to save the `_ID` field(unless something in the provider has injected it). I'll look into that.

**Comment by stackcoder** [comment]
Hey @dmfs,
are there any news to this? This bug is really annoying and unfortunately it's not easy for me to get into this code to find a solution according to the concept here.

**Comment by dmfs** [comment]
Unfortunately no. I haven't been able to reproduce it so far and I have no idea on how the `_ID` value managed to get into the data set.

Which version do you use? Does it happen every time? Does it only happen when you update the due date or does it also happen when you update other fields?

**Comment by mikebabb** [comment]
Just started having this problem too. I assumed it was because of the update to Android 10 yesterday, but it seems not as this issue is about a month old.

I reported it and sent a log via the "app keeps crashing" dialogue in Android itself - does that send the info you need, @dmfs?

**Comment by mikebabb** [comment]
FWIW, it only happens to me on editing the due date or time of an existing task. Editing other fields like the title works without issue.

**Comment by dmfs** [comment]
@mikebabb I'll check the play store reports for this, thanks.

I'm going to release OpenTasks 1.2 during the next couple of days and I'll have a look into this afterwards.

**Comment by mikebabb** [comment]
Let me know if I can provide any more information to help.

**Comment by anirudhnihalani** [comment]
I am also having a similar issue, changing due date of an existing task either crashes the app and loses all changes or just doesn't save the changes without crashing. 

Also if start date is selected to be beyond due date the app crashes instead of giving any warning.

**Comment by aubrel** [comment]
Hey, just throwing my hat into the ring as well, the exact same issue is happening to me, and I also thought it was due to the upgrade to Android 10 since it had never happened before, but again, apparently not. I also am having the same issue as @mikebabb , in that every other field is fine save for the due date field.

**Comment by dmfs** [comment]
I haven't been able to reproduce it yet. Does it work for you when you change the due date via the `+1d` or `+1h` button in the details view?
I'm wondering whether this is a provider or a UI issue.

**Comment by faheinz** [comment]
@dmfs - For tasks that already have dates, I am able to change the due date/time using the +1hr and +1 day buttons, however this only works (of course) for tasks that already have due dates assigned to them. For tasks that have not yet had dates assigned, what happens is:

- I open the task intending to edit the due date.
- I select a due date.
- I attempt to save my changes, and am returned to viewing the task, almost as though they had been saved, but there is no date information shown.
- I attempt to change the date again, and then the app crashes. There is no ability to use the +1hr or day buttons, since no date has actually been saved.

Some info about my device:

Model: Pixel 3 (not rooted)
Build no.: QP1A.190711.020
OS: Android 10
App version: 1.1.20 from F-Droid

**Comment by aubrel** [comment]
@dmfs - For tasks that already have dates, I am able to successfully change the due date/time using the +1hr and +1 day buttons, however this only works (of course) for tasks that already have due dates assigned to them. For tasks that have not yet had dates assigned, what happens is:

- I open the task intending to edit the due date.
- I select a due date.
- I attempt to save my changes, and am returned to viewing the task, almost as though they had been saved, but there is no date information shown.
- I attempt to change the date again, but upon clicking "Save," the app crashes. There is no ability to use the +1hr or day buttons, since no date has actually been saved.

Some info about my device:

Model: Pixel 3 (not rooted)
Build no.: QP1A.190711.020
OS: Android 10
App version: 1.1.20 from F-Droid

**Comment by mikebabb** [comment]
I'm on version 1.2.1 of OpenTasks, and interestingly having just updated Tasks on my Nextcloud instance to version 0.11.3 I'm no longer having the problem. I can now add a due date to existing tasks that don't have one, or edit tasks that already have a due date and assign them a new date without the app crashing.

Any Nextcloud users still having this issue, check whether there's an update to the Tasks app available. I was only notified of the update today but it may have solved this problem for me.

**Comment by dmfs** [comment]
I still see (relatively few) reports of this on Google Play, so it's not fixed by 1.2.1. On the other hand, I still can't reproduce it myself. Not on Android 9 nor on Android 10.

**Comment by aubrel** [comment]
A little more detail that I just noticed today, not sure if this is useful, especially if you haven't been able to reproduce it; but, I am also noticing that, even after refreshing, the application does not seem to reflect changes made on the server side.

For example, I used another app to update a due date, and confirmed that the change was made on the server using Thunderbird. I then ensured to refresh CalDav on my phone, but the new changes still did not appear.

So, not sure if that's related or not, but since it's happening on a device with this due date crash issue too, maybe it will be. There it is anyway.
